rmf_api_msgs/samples/task_state/multi_dropoff_delivery.json

```diff
@@ -0,0 +1,214 @@
+          "detail": {
+            "location": "Kitchen",
+            "items": [
+              {
+                "type": "soda",
+                "quantity": 1
+              },
+              {
+                "type": "water",
+                "quantity": 1
+              }
+            ]
+          },
+          "estimate_millis": 600000
+        },
+        "final_event_id": 0,
+        "events": [
+          {
+            "id": 0,
+            "status": "completed",
+            "name": "Sequence",
+            "detail": "",
+            "log": [],
+            "deps": [1, 2]
+          },
+          {
+            "id": 1,
+            "status": "completed",
+            "name": "Go to [place:kitchen]",
+            "detail": "",
+            "log": [
+              {
+                "tier": "info",
+                "unix_millis_time": 1636388415000,
+                "text": "Waiting for [door:kitchen_door] to open"
+              },
+              {
+                "tier": "info",
+                "unix_millis_time": 1636388422000,
+                "text": "Passed through [door:kitchen_door]"
+              },
+              {
+                "tier": "info",
+                "unix_millis_time": 1636388426000,
+                "text": "Arrived at [place:kitchen]"
+              }
+            ],
+            "deps": []
+          },
+          {
+            "id": 2,
+            "status": "completed",
+            "name": "Receive items",
+            "detail": [
+              {
+                "type": "soda",
+                "quantity": 1
+              },
+              {
+                "type": "water",
+                "quantity": 1
+              }
+            ],
+            "log": [
+              {
+                "tier": "info",
+                "unix_millis_time": 1636388430000,
+                "text": "Received [item:soda]"
+              },
+              {
+                "tier": "info",
+                "unix_millis_time": 1636388440000,
+                "text": "Received [item:water]"
+              }
+            ],
+            "deps": []
+          }
+        ]
+      }
+    }
+  ],
+  "active": {
+    "id": 2,
+    "header": {
+      "category": "Drop Off",
+      "detail": {
+        "location": "room_203",
+        "items": [
+          {
+            "type": "soda",
+            "quantity": 1
+          }
+        ]
+      },
+      "estimate_millis": 720000
+    },
+    "final_event_id": 0,
+    "events": [
+      {
+        "id": 0,
+        "status": "underway",
+        "name": "Sequence",
+        "detail": "",
+        "log": [],
+        "deps": [1, 2]
+      },
+      {
+        "id": 1,
+        "status": "underway",
+        "name": "Go to [place:room_203]",
+        "detail": "",
+        "log": [
+          {
+            "tier": "info",
+            "unix_millis_time": 1636388445000,
+            "text": "Waiting for [door:kitchen_door] to open"
+          }
+        ],
```

Comment:
It seems like `log` could be pretty big, should be move this to a separate schema? A common way is to have it be a string representing an url.

Reply:
`log_entry` does have its own schema; this file is a sample of the schema being applied.

Reply:
Sorry, I meant to move this outside events etc and be its own top level schema, but I guess it depends on how often we expect to see task state updates and how much logs will it stack up overtime. If it's a relatively small amount then it's fine as it is.

Reply:
I see what you mean.

One thing I've personally found bothersome with RMF, as someone who needs to monitor what's going on in the system while it's running, is how the output on the terminal is just a dump of little reports from every single component at once. Sometimes I'd really like to ask "What's going on with this event (e.g. the robot waiting to enter a lift)?" and just be able to see a record of "(1) The robot has requested the lift. (2) The robot is waiting for lift to become available. (3) The lift is now available. (4) The lift is now coming to pick up the robot. ... etc ...".

The rosconsole does allow some degree of filtering, but to make that useful we'd need to design and remember special tokens to shove into every log message. I thought this structure would be nicer because it tells the dashboard exactly what system components to associate the log message with, and dashboard can use that to make the UI better at delivering the information that the user wants. For example, we could make it so that clicking on the event element in the UI will fill in a panel with the log information specific to that event.

For storage, we could copy and funnel all these log messages into a single database where key the entries based on `(task ID, phase ID, event ID)`.

Reply:
It only becomes an issue if tasks builds up a large amount of logs and task states are being updated frequently. If the logs are separated, we don't need to download that data until the user clicks on an event, task state updates also does not need to contain all the logs.

This is probably going to be a much bigger problem for robot state, those are expected to be updated multiple times per second and would quickly consume lots of bandwidth.

Reply:
Ah right, so I should clarify something about these schemas which isn't written down anywhere yet: The messages will not always contain all the fields. That's why none of the fields are marked as required (although I'll probably make `id` required, because that one really is always needed). There are only two situations where each field will be included:
1. When the listener first subscribes and is receiving their first state update message
2. When the value of the field has changed since the last time an update was sent

When it comes to lists, some lists will have `append` behavior while others will have `replace` behavior, which indicates how the update for that list should be handled.

I just noticed your edit after typing out the above, and you're exactly right.

> this does not address the problem that the logs will eventually build up and it's bad for the client to download everything

The dashboard can safely clear out the logs as each task is finished. The express purpose of these log fields is to provide information that the lower level system believes will be relevant to a human operator who is actively monitoring the progress of the tasks. The expectation is that it will only include notable events that relate to progress milestones or important warnings/errors that an operator should be made aware of. For lower level highly detailed system logs, we can have a different logging pipeline that isn't fed directly into the dashboard.

Reply:
> The dashboard can safely clear out the logs as each task is finished. The express purpose of these log fields is to provide information that the lower level system believes will be relevant to a human operator who is actively monitoring the progress of the tasks.

We need to display a history of tasks and displaying these logs would be desired. `robot_state` also a `log` field and clearing this field is not as straight forward as the task logs.

Reply:
I think the problem of "How much history should we hold" is something we need to contend with no matter what. For example: How many completed tasks should remain in the dashboard display at any given time?

I can think of a few possible solutions, but we'll probably want to discuss the tradeoffs offline:
1. Keep a list of completed tasks but clear out the state information related to that task. If a user asks to view the state information of a completed task, then the dashboard queries a database to retrieve that information and clears it out again some time later.
2. Keep the state information of the last N completed tasks, and use (1) if the user asks to see the state of a completed task whose state has already been cleared.
3. Some cross between (1) and (2) where we keep a limited-size cache of whichever completed tasks the user has most recently asked to view.

As for the `robot_state` log, I imagine we'll need to do something similar but at the log level. E.g. keep the last N log entries on display, and when the user wants to see logs from further back maybe we have them query a database for some time range and then display the logs for that range in a pop up screen. When they close the screen we clear those logs out of the dashboard.